src/analysis/calculus/cont_diff.lean

```diff
@@ -2308,44 +2324,171 @@ Warning: see remarks attached to `cont_diff_prod_assoc`
-      cont_diff_within_at_succ_iff_has_fderiv_within_at'.mp (hf.of_le hkn),
-    apply hf'.congr_of_eventually_eq_insert,
-    filter_upwards [hv, hs],
+      (hf.of_le $ (add_le_add_right hkm 1).trans hmn).has_fderiv_within_at_nhds (hg.of_le hkm)
+      hst hu,
+    refine hf'.congr_of_eventually_eq_insert _,
+    filter_upwards [hv, ht],
     exact λ y hy h2y, (hvf' y hy).fderiv_within h2y },
   induction m using with_top.rec_top_coe,
   { obtain rfl := eq_top_iff.mpr hmn,
     rw [cont_diff_within_at_top],
     exact λ m, this m le_top },
-  exact this m hmn
-end
-
-lemma cont_diff_within_at.fderiv_within
+  exact this m le_rfl
+end
+
+/-- A special case of `cont_diff_within_at.fderiv_within''` where we require that
+  `s ∪ {x} ⊆ g⁻¹(t)`. -/
+lemma cont_diff_within_at.fderiv_within' {f : E → F → G} {g : E → F} {u : set (E × F)}
+  {t : set F} {n : ℕ∞}
+  (hf : cont_diff_within_at 𝕜 n (function.uncurry f) u (x, g x))
+  (hg : cont_diff_within_at 𝕜 m g s x)
+  (ht : ∀ᶠ x' in 𝓝[insert x s] x, unique_diff_within_at 𝕜 t (g x'))
+  (hmn : m + 1 ≤ n)
+  (hst : insert x s ×ˢ t ⊆ u)
+  (h2st : s ⊆ g ⁻¹' t) :
+  cont_diff_within_at 𝕜 m (λ x, fderiv_within 𝕜 (f x) t (g x)) s x :=
+begin
+  refine hf.fderiv_within'' hg ht hmn hst _,
+  refine mem_of_superset self_mem_nhds_within _,
+  refine image_prod_mk_subset_prod.trans _,
+  rw [image_id'], rw [← image_subset_iff] at h2st,
+  exact (prod_mono (subset_insert x s) h2st).trans hst
+end
+
+/-- A special case of `cont_diff_within_at.fderiv_within'` where we require that `x ∈ s` and there
+  are unique derivatives everywhere within `t`. -/
+lemma cont_diff_within_at.fderiv_within {f : E → F → G} {g : E → F} {u : set (E × F)}
+  {t : set F} {n : ℕ∞}
+  (hf : cont_diff_within_at 𝕜 n (function.uncurry f) u (x, g x))
+  (hg : cont_diff_within_at 𝕜 m g s x)
+  (ht : unique_diff_on 𝕜 t)
+  (hmn : m + 1 ≤ n) (hx : x ∈ s)
+  (hst : s ×ˢ t ⊆ u)
+  (h2st : s ⊆ g ⁻¹' t) :
+  cont_diff_within_at 𝕜 m (λ x, fderiv_within 𝕜 (f x) t (g x)) s x :=
+begin
+  rw [← insert_eq_self.mpr hx] at hst,
+  refine hf.fderiv_within' hg _ hmn hst h2st,
+  rw [insert_eq_self.mpr hx],
+  exact eventually_of_mem self_mem_nhds_within (λ x hx, ht _ (h2st hx))
+end
+
+/-- `fderiv` applied to a (variable) vector is smooth at a point within a set. -/
+lemma cont_diff_within_at.fderiv_within_apply {f : E → F → G} {g k : E → F} {u : set (E × F)}
+  {t : set F} {n : ℕ∞}
+  (hf : cont_diff_within_at 𝕜 n (function.uncurry f) u (x, g x))
+  (hg : cont_diff_within_at 𝕜 m g s x)
+  (hk : cont_diff_within_at 𝕜 m k s x)
+  (ht : unique_diff_on 𝕜 t)
+  (hmn : m + 1 ≤ n) (hx : x ∈ s)
+  (hst : s ×ˢ t ⊆ u)
+  (h2st : s ⊆ g ⁻¹' t) :
+  cont_diff_within_at 𝕜 m (λ x, fderiv_within 𝕜 (f x) t (g x) (k x)) s x :=
+(cont_diff_fst.clm_apply cont_diff_snd).cont_diff_at.comp_cont_diff_within_at x
+  ((hf.fderiv_within hg ht hmn hx hst h2st).prod hk)
+
+/-- `fderiv_within` is smooth at `x` within `s` (for functions without parameters). -/
```

Comment:
```suggestion
/-- `fderiv_within 𝕜 f s` is smooth at `x` within `s` (for functions without parameters). -/
```
(Otherwise, one may think that you are talking of the function `fderiv_within` itself, from spaces of functions from spaces of functions. Same thing in all the docstrings).